### PR TITLE
Integrate Allocation Dispatcher into the memory management system

### PR DESCRIPTION
**Pull request by Scanteianu**
Replace calls to malloc with a bespoke allocator which is optimized for allocating objects of a finite set of sizes

### CODE REVIEW

File: include/allocationDispatcher.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
+#ifndef ALLOCATIONDISPATCHER
```
Comment (j-hui) [inline include/allocationDispatcher.h]: Change header guard to `_ALLOCATION_DISPATCHER_H`

File: include/allocationDispatcher.h [hunk 2]
```diff
@@ -0,0 +1,70 @@
+#ifndef ALLOCATIONDISPATCHER
+#define ALLOCATIONDISPATCHER
+#include "fixedAllocator.h"
+#include "constants.h"
+#include <stddef.h>
+#include <stdio.h>
```
Comment (j-hui) [inline include/allocationDispatcher.h]: should not be necessary

File: include/constants.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#ifndef constants
+#define constants
+#include <stdint.h>
+// inorder to facilitate, align any addresses to word boundary if malloc doesn't
+#if UINTPTR_MAX == 0xFFFF
```
Comment (j-hui) [inline include/constants.h]: This is already defined in `ssm.h`. No need to redefine

File: include/constants.h [hunk 4]
```diff
@@ -0,0 +1,21 @@
+#ifndef constants
```
Comment (j-hui) [inline include/constants.h]: this file doesn't seem necessary. `sizeof(ssm_word_t)` is more self-explanatory inline and does not take up that much more space, and the debug stuff should be defined on a per-compilation unit basis

File: include/fixedAllocator.h [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+#include <stdint.h>
+
+// initialize func allocates space to a global var
+// skip list initialized with first word pointing to next free
+// global next free pointer - head
+// when freeing, insert at head of linkedlist (reuse!, reduce!, recycle!)
+// actually - not a global var - i guess it's a struct called the fixed_allocator
+// this struct holds memory, and a fixed size
+
+typedef void* MEMORY;
```
Comment (j-hui) [inline include/fixedAllocator.h]: This hardly seems necessary

File: include/fixedAllocator.h [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+#define fixed_allocator
+
+#include <stddef.h>
+#include <stdint.h>
+
+// initialize func allocates space to a global var
+// skip list initialized with first word pointing to next free
+// global next free pointer - head
+// when freeing, insert at head of linkedlist (reuse!, reduce!, recycle!)
+// actually - not a global var - i guess it's a struct called the fixed_allocator
```
Comment (j-hui) [inline include/fixedAllocator.h]: are these notes, or documentation? Please tidy this up

File: src/allocationDispatcher.c [hunk 7]
```diff
@@ -0,0 +1,55 @@
+#include "allocationDispatcher.h"
+#include <stdlib.h>
+#include <stdio.h>
+#include <assert.h>
+
+ALLOCATION_DISPATCHER* adInitialize(size_t blockSizes[], size_t numBlocks[], size_t numAllocators, void* memoryPool){
+  void* memoryHead = memoryPool;
+  ALLOCATION_DISPATCHER* dispatcher= memoryHead;
+  memoryHead = ((char*) memoryHead)+sizeof(ALLOCATION_DISPATCHER);
```
Comment (j-hui) [inline src/allocationDispatcher.c]: no need to cast to `char *`; `void *` pointer arithmetic should work as expected.

File: include/allocationDispatcher.h [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+  The child allocators make fixed-size allocations, and the dispatcher
+  will select the apropriate child allocator.
+*/
+
+/*
+  adInitialize:
+
+  adInitialize sets up the dispatcher, initializing underlying fixed allocators
+*/
+ALLOCATION_DISPATCHER* adInitialize(size_t blockSizes[], size_t numBlocks[], size_t numAllocators, void* memoryPool);
```
Comment (j-hui) [inline include/allocationDispatcher.h]: What do the parameters mean? Please document.

File: src/ssm-mem.c [hunk 9]
```diff
@@ -4,13 +4,43 @@
...
  */
 #include <ssm-internal.h>
-
+#include <allocationDispatcher.h>
+ALLOCATION_DISPATCHER *ad = NULL;
+size_t allSizes[1000];
+int numSizes = 0;
 struct ssm_mm *ssm_mem_alloc(size_t size) {
-  return malloc(size); // TODO: (dan) use our own allocator
+  if(ad==NULL){
```
Comment (j-hui) [inline src/ssm-mem.c]: Adding this check every single time is unnecessarily costly. Initialization should be performed in a separate function. Add an `ssm_mem_init()` function for that.

File: src/ssm-mem.c [hunk 10]
```diff
@@ -4,13 +4,43 @@
...
+    size_t allocatorSizes[] = {48};
+    size_t allocatorBlocks[] = {4000};
+    size_t numAllocators = 1;
+    void* memoryBlock =  malloc(48*1000*WORD_BOUNDARY);
```
Comment (j-hui) [inline src/ssm-mem.c]: By default, this should not malloc. Instead, use a static memory pool.

File: src/ssm-mem.c [hunk 11]
```diff
@@ -4,13 +4,43 @@
  *  @author John Hui (j-hui)
  */
 #include <ssm-internal.h>
-
+#include <allocationDispatcher.h>
+ALLOCATION_DISPATCHER *ad = NULL;
+size_t allSizes[1000];
```
Comment (j-hui) [inline src/ssm-mem.c]: Why is this global? What is this?

File: src/allocationDispatcher.c [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    #if !ALLOW_MALLOC
+      assert(0);
+    #endif
+    free(memory);
+  }
+  else{
+    faFree(allocator,memory);
+  }
+}
+FIXED_ALLOCATOR* findAllocator(ALLOCATION_DISPATCHER *ad, size_t size){
```
Comment (j-hui) [inline src/allocationDispatcher.c]: Please fix this to work in constant time.

File: src/allocationDispatcher.c [hunk 13]
```diff
@@ -0,0 +1,55 @@
+#include "allocationDispatcher.h"
+#include <stdlib.h>
+#include <stdio.h>
+#include <assert.h>
+
+ALLOCATION_DISPATCHER* adInitialize(size_t blockSizes[], size_t numBlocks[], size_t numAllocators, void* memoryPool){
```
Comment (j-hui) [inline src/allocationDispatcher.c]: I can barely understand what this code is doing. Why is `memoryHead` being assigned multiple times? What does the variable name mean? If it has a fixed memory layout, then it should be cast to an appropriately defined struct expressing as much.

File: src/ssm-mem.c [hunk 14]
```diff
@@ -4,13 +4,43 @@
...
+    void* memoryBlock =  malloc(48*1000*WORD_BOUNDARY);
+    ALLOCATION_DISPATCHER *dispatcher = adInitialize(allocatorSizes,allocatorBlocks,numAllocators,memoryBlock);
+    ad = dispatcher;
+  }
+  /*
+
+  // this block of code is used to (inefficiently) construct an array of all allocated sizes
+
+  int foundSize = 0;
+  for(int i=0; i<numSizes;i++){
```
Comment (j-hui) [inline src/ssm-mem.c]: Please remove this dead code, unless this is meant to be integrated.

File: include/fixedAllocator.h [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+
+typedef void* MEMORY;
+
+
+#if UINTPTR_MAX == 0xFFFF
+  #error how do you even metadata?
+#elif INTPTR_MAX == INT32_MAX
+  typedef uint32_t INTEGER_REPRESENTATION;
+#elif INTPTR_MAX == INT64_MAX
+  typedef unsigned long INTEGER_REPRESENTATION;
```
Comment (j-hui) [inline include/fixedAllocator.h]: Why is this needed if you already have `ssm_word_t`?

**Review by j-hui — CHANGES_REQUESTED** [review]
Please follow naming conventions of the rest of this library, including:
- snake case for identifiers
- `_t` suffix for typedefs
- hyphen-case (aka kebab case) for filenames
- `_HEADER_FILENAME_H` for header guards
- Doxygen doc comments

Everything under `include/` is under the include path, so you should just import using `#include <file.h>` instead of with quotation marks.

It's not clear why allocationDispatcher and fixedAllocator need to be in separate compilation units; they aren't very large, and can readably fit `ssm-mem.c`.

Also, please run some kind of formatter over your code; I suggest `clang-format`.

File: src/ssm-mem.c (hunk 11 above)
Comment (Scanteianu) [reply src/ssm-mem.c]: no longer needed

File: src/allocationDispatcher.c (hunk 13 above)
Comment (Scanteianu) [reply src/allocationDispatcher.c]: this is kind of difficult, as the memory layout isn't quite fixed - each underlying allocator can be a different size. Added a comment explaining what is going on.

Comment (j-hui) [reply src/allocationDispatcher.c]: Then this is our chance to pick a sensible memory layout.

I'm even not really sure it makes sense to make this as configurable as suggested by this interface. For instance, it makes no sense initialize the the allocation pool with sizes `{1, 2, 7, 3, 5, 49}` or something crazy like that. I'm not even convinced there's much benefit in configuring this differently at runtime.

@sedwards-lab do you have any thoughts on this? Are there applications where we would dramatically benefit from having different block size configurations?

Comment (j-hui) [reply src/allocationDispatcher.c]: Also, we really should be using an on-demand paging mechanism, so that the underlying allocator can request new pages dynamically. We should also fix the page size to make things easier, and maybe just have allocator pools for blocks of size 4, 16, 64, 256, and 1024 (or something like that). That would suggest an interface like:

```
allocation_dispatcher_t *ad_initialize(/* other config */ void *(*get_page)(size_t));
```

where `get_page` is a callback that the dispatcher can use to request more pages to expand its memory pool; the `size_t` parameter would be used to indicate how many pages worth of memory are required. For instance, on UNIX-like systems, `get_page` could use the `mmap` syscall to feed more pages to each allocator.

**Comment by j-hui** [comment]
Superseded by #6
